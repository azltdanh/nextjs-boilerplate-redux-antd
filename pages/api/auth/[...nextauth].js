import NextAuth from 'next-auth';
import Providers from 'next-auth/providers';

const options = {
  // pages: {
  //   signIn: '/auth/signin',
  //   signOut: '/auth/signout',
  //   error: '/auth/error', // Error code passed in query string as ?error=
  //   verifyRequest: '/auth/verify-request', // (used for check email message)
  //   newUser: null // If set, new users will be directed here on first sign in
  // },
  providers: [
    // OAuth authentication providers...
    Providers.Facebook({
      clientId: process.env.FACEBOOK_ID,
      clientSecret: process.env.FACEBOOK_SECRET,
    }),
    Providers.Google({
      clientId: process.env.GOOGLE_ID,
      clientSecret: process.env.GOOGLE_SECRET,
    }),
    // Password-less / email sign in
    // Providers.Email({
    //   server: process.env.MAIL_SERVER,
    //   from: process.env.MAIL_FROM,
    // }),
  ],

  // Optional SQL or MongoDB database to persist users
  // database: process.env.DATABASE_URL,
};

export default (req, res) => NextAuth(req, res, options);
