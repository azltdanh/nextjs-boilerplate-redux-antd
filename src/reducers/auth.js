import { HYDRATE } from 'next-redux-wrapper';

export const AUTH = {
  // SIGN_IN
  SIGN_IN_REQUEST: 'USER/SIGN_IN_REQUEST',
  SIGN_IN_REQUEST_SUCCESS: 'USER/SIGN_IN_REQUEST_SUCCESS',
  SIGN_IN_REQUEST_FAILURE: 'USER/SIGN_IN_REQUEST_FAILURE',
  // SIGN_OUT
  SIGN_OUT_REQUEST: 'USER/SIGN_OUT_REQUEST',
  SIGN_OUT_REQUEST_SUCCESS: 'USER/SIGN_OUT_REQUEST_SUCCESS',
  SIGN_OUT_REQUEST_FAILURE: 'USER/SIGN_OUT_REQUEST_FAILURE',
  // SIGN_UP
  SIGN_UP_REQUEST: 'USER/SIGN_UP_REQUEST',
  SIGN_UP_REQUEST_SUCCESS: 'USER/SIGN_UP_REQUEST_SUCCESS',
  SIGN_UP_REQUEST_FAILURE: 'USER/SIGN_UP_REQUEST_FAILURE',
};

export const signIn = (payload) => ({
  type: AUTH.SIGN_IN_REQUEST,
  payload,
});

export const signOut = () => ({
  type: AUTH.SIGN_OUT_REQUEST,
});

export const signUp = (payload) => ({
  type: AUTH.SIGN_UP_REQUEST,
  payload,
});

export const initialState = {
  loading: false,
  data: null,
  error: null,
};

const auth = (state = initialState, action = {}) => {
  switch (action.type) {
    case HYDRATE: {
      return {
        ...state,
        ...action.payload.auth, // pick the 'auth' only
      };
    }
    case AUTH.SIGN_IN_REQUEST:
    case AUTH.SIGN_OUT_REQUEST:
    case AUTH.SIGN_UP_REQUEST: {
      return {
        ...state,
        loading: true,
      };
    }
    case AUTH.SIGN_IN_REQUEST_SUCCESS:
    case AUTH.SIGN_UP_REQUEST_SUCCESS: {
      return {
        ...state,
        loading: false,
        data: action.payload,
      };
    }
    case AUTH.SIGN_OUT_REQUEST_SUCCESS: {
      return {
        ...initialState,
      };
    }
    case AUTH.SIGN_IN_REQUEST_FAILURE:
    case AUTH.SIGN_OUT_REQUEST_FAILURE:
    case AUTH.SIGN_UP_REQUEST_FAILURE: {
      return {
        ...state,
        loading: false,
        error: action.payload,
      };
    }
    default: {
      return { ...state };
    }
  }
};

export default auth;
