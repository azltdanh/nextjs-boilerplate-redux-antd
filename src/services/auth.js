import api from './api';

export const signIn = async ({ email, password }) => {
  const resp = await api.get(`/users?email=${email}&password=${password}`);
  return resp.data[0];
};

export const signOut = async (payload) => {
  const resp = await api.post('/sign-out', payload).catch(() => ({})); // mock api
  return resp.data;
};

export const signUp = async (payload) => {
  const resp = await api.post('/users', payload);
  return resp.data;
};
