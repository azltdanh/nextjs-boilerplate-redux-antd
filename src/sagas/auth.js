import { all, call, put, takeLatest } from 'redux-saga/effects';
import { signIn, signOut, signUp } from '@services/auth';
import { AUTH } from '@reducers/auth';

export function* signInSaga(action) {
  try {
    const data = yield call(signIn, action.payload);
    yield put({ type: AUTH.SIGN_IN_REQUEST_SUCCESS, payload: data });
  } catch (error) {
    yield put({ type: AUTH.SIGN_IN_REQUEST_FAILURE, payload: error });
  }
}

export function* signOutSaga(action) {
  try {
    const data = yield call(signOut, action.payload);
    yield put({ type: AUTH.SIGN_OUT_REQUEST_SUCCESS, payload: data });
  } catch (error) {
    yield put({ type: AUTH.SIGN_OUT_REQUEST_FAILURE, payload: error });
  }
}

export function* signUpSaga(action) {
  try {
    const data = yield call(signUp, action.payload);
    yield put({ type: AUTH.SIGN_UP_REQUEST_SUCCESS, payload: data });
  } catch (error) {
    yield put({ type: AUTH.SIGN_UP_REQUEST_FAILURE, payload: error });
  }
}

function* authSagas() {
  yield all([
    takeLatest(AUTH.SIGN_IN_REQUEST, signInSaga),
    takeLatest(AUTH.SIGN_OUT_REQUEST, signOutSaga),
    takeLatest(AUTH.SIGN_UP_REQUEST, signUpSaga),
  ]);
}

export default authSagas;
