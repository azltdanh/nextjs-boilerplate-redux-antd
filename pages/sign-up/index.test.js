import { useSelector } from 'react-redux';
import { initialState, signUp } from '@reducers/auth';
import mockUser from '@mock/user';
import { Form } from 'antd';
import SignUpPage from '.';

// mock router history
const mockHistoryPush = jest.fn();
jest.mock('next/router', () => ({
  ...jest.requireActual('next/router'),
  useRouter: () => ({
    pathname: '/',
    push: mockHistoryPush,
  }),
}));

// mock redux store
const mockUseSelector = (mockedStore) => {
  useSelector.mockImplementation((callback) => callback(mockedStore));
};
const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

// mock timers
jest.useFakeTimers();

describe(SignUpPage.name, () => {
  let wrapper;

  beforeEach(() => {
    mockDispatch.mockClear();
    const mockStore = { auth: { ...initialState } };
    mockUseSelector(mockStore);
    wrapper = mount(<SignUpPage />);
  });

  it('renders', () => {
    expect(shallow(<SignUpPage />)).toMatchSnapshot();
  });

  it('dispatch signUp success', () => {
    const formValues = {
      fullname: mockUser.fullname,
      prefix: mockUser.prefix,
      phone: mockUser.phone,
      email: mockUser.email,
      password: '1',
      confirm: '1',
    };
    const appForm = wrapper.find(Form);
    appForm.props().form.setFieldsValue({ ...formValues });

    wrapper
      .find('input#fullname')
      .simulate('change', { target: { value: mockUser.fullname } });
    wrapper
      .find('input#phone')
      .simulate('change', { target: { value: mockUser.phone } });
    wrapper
      .find('input#email')
      .simulate('change', { target: { value: mockUser.email } });
    wrapper
      .find('input#password')
      .simulate('change', { target: { value: '1' } });
    wrapper
      .find('input#confirm')
      .simulate('change', { target: { value: '2' } }) // trigger validate
      .simulate('change', { target: { value: '1' } });
    wrapper.find('form').simulate('submit');
    appForm.props().onFinish(formValues);

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(mockDispatch).toHaveBeenCalledWith(signUp(formValues));
  });

  it('redirect when user is signed in', () => {
    const mockStore = { auth: { ...initialState, data: { ...mockUser } } };
    mockUseSelector(mockStore);
    mount(<SignUpPage />);
    expect(mockHistoryPush).toHaveBeenCalledWith('/');
  });
});
