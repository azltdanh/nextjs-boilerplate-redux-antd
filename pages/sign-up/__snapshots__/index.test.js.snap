// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpPage renders 1`] = `
<ForwardRef(InternalForm)
  className="app-form"
  form={
    Object {
      "__INTERNAL__": Object {
        "itemRef": [Function],
      },
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "getInternalHooks": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "resetFields": [Function],
      "scrollToField": [Function],
      "setFields": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
    }
  }
  initialValues={
    Object {
      "prefix": "1",
    }
  }
  labelCol={
    Object {
      "sm": Object {
        "span": 8,
      },
      "xs": Object {
        "span": 0,
      },
    }
  }
  onFinish={[Function]}
  requiredMark={false}
  scrollToFirstError={true}
  wrapperCol={
    Object {
      "sm": Object {
        "span": 16,
      },
      "xs": Object {
        "span": 24,
      },
    }
  }
>
  <FormItem
    label="Full name"
    name="fullname"
    rules={
      Array [
        Object {
          "message": "Please input your full name!",
          "required": true,
        },
      ]
    }
  >
    <Input
      placeholder="Full name"
      type="text"
    />
  </FormItem>
  <FormItem
    label="Phone number"
    name="phone"
    rules={
      Array [
        Object {
          "message": "Please input your phone number!",
          "required": true,
        },
      ]
    }
  >
    <Input
      addonBefore={
        <FormItem
          name="prefix"
          noStyle={true}
        >
          <Select
            bordered={true}
            choiceTransitionName=""
            style={
              Object {
                "width": 64,
              }
            }
            transitionName="slide-up"
          >
            <Option
              value="1"
            >
              +1
            </Option>
            <Option
              value="84"
            >
              +84
            </Option>
          </Select>
        </FormItem>
      }
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
    />
  </FormItem>
  <FormItem
    label="Email"
    name="email"
    rules={
      Array [
        Object {
          "message": "Please input your valid email!",
          "required": true,
          "type": "email",
        },
      ]
    }
  >
    <Input
      placeholder="Email"
      type="text"
    />
  </FormItem>
  <FormItem
    hasFeedback={true}
    label="Password"
    name="password"
    rules={
      Array [
        Object {
          "message": "Please input your password!",
          "required": true,
        },
      ]
    }
  >
    <Password
      action="click"
      iconRender={[Function]}
      placeholder="Password"
      visibilityToggle={true}
    />
  </FormItem>
  <FormItem
    dependencies={
      Array [
        "password",
      ]
    }
    hasFeedback={true}
    label="Confirm password"
    name="confirm"
    rules={
      Array [
        Object {
          "message": "Please confirm your password!",
          "required": true,
        },
        [Function],
      ]
    }
  >
    <Password
      action="click"
      iconRender={[Function]}
      placeholder="Confirm password"
      visibilityToggle={true}
    />
  </FormItem>
  <FormItem
    className="btn-wrapper"
    wrapperCol={
      Object {
        "sm": Object {
          "offset": 8,
          "span": 16,
        },
        "xs": Object {
          "offset": 0,
          "span": 24,
        },
      }
    }
  >
    <ButtonSubmit
      loading={false}
    >
      Sign Up
    </ButtonSubmit>
  </FormItem>
</ForwardRef(InternalForm)>
`;
